lightning/restore/restore.go

```diff
@@ -306,15 +307,36 @@ func (rc *RestoreController) restoreSchema(ctx context.Context) error {
 			defer db.Close()
 			db.ExecContext(ctx, "SET SQL_MODE = ?", rc.cfg.TiDB.StrSQLMode)
 		}
-
+		// CONFUSED: how we control concurrency via configuration?
+		// OR do not implement glue.SQLExecutor inside ExternalTiDBGlue
+		// Maybe we can aggregate an instance of glue.SQLExecutor?
+		concurrency := 16
+		var sqlExec glue.SQLExecutor
+		var err error
+		if concurrency > 1 {
+			dbFactory := func(conf config.DBStore) (*sql.DB, error) {
+				db, err := DBFromConfig(conf)
+				if err == nil {
+					db.ExecContext(ctx, "SET SQL_MODE = ?", conf.StrSQLMode)
+				}
+				return db, err
+			}
+			sqlExec, err = glue.NewDBPool(concurrency, rc.cfg.TiDB, dbFactory)
+			defer sqlExec.Close()
+			if err != nil {
+				return errors.Trace(err)
+			}
+		} else {
+			sqlExec = rc.tidbGlue.GetSQLExecutor()
+		}
 		for _, dbMeta := range rc.dbMetas {
 			task := log.With(zap.String("db", dbMeta.Name)).Begin(zap.InfoLevel, "restore table schema")
 
 			tablesSchema := make(map[string]string)
 			for _, tblMeta := range dbMeta.Tables {
 				tablesSchema[tblMeta.Name] = tblMeta.GetSchema(ctx, rc.store)
 			}
-			err := InitSchema(ctx, rc.tidbGlue, dbMeta.Name, tablesSchema)
+			err = InitSchema(ctx, concurrency, rc.tidbGlue.GetParser(), sqlExec, dbMeta.Name, tablesSchema)
```

Comment:
I think it's better to run these create stmts in following order:
1. create all databases in parallel
2. create all tables in parallel
3. create all views in parallel

or an alternative choices is:
spawn a pool of executor routines. then first send all the create db tasks to the executor task chan. Then wait resp from result chan, after receive a finishing create db msg, pipeline add all create table task for the db into the task chan. After create all db tasks finished for one db, add all create view task then... wait all tasks to be finished.

@lance6716 @kennytm  what do you think

Reply:
@glorv It is do a better way, we can control those steps like a finite-state machine.